src/main/java/org/opensearch/neuralsearch/plugin/query/NeuralQueryBuilder.java

```diff
@@ -172,10 +193,35 @@ private static void parseQueryParams(XContentParser parser, NeuralQueryBuilder n
         }
     }
 
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) {
+        // When re-writing a QueryBuilder, if the QueryBuilder is not changed, doRewrite should return itself.
+        // Otherwise, it should return the modified copy. With the asynchronous call, on first rewrite, we create a new
+        // vector supplier that will get populated once the asynchronous call finishes and pass this supplier in to
+        // create a new builder. Once the supplier's value gets set, we return a KNNQueryBuilder. Otherwise, we just
+        // return the current unmodified query builder.
+        if (vectorSupplier() != null) {
+            return vectorSupplier().get() == null ? this : new KNNQueryBuilder(fieldName(), vectorSupplier.get(), k());
+        }
+
+        if (vectorSupplier() == null) {
+            SetOnce<float[]> vectorSetOnce = new SetOnce<>();
+            queryRewriteContext.registerAsyncAction(
+                ((client, actionListener) -> ML_CLIENT.inferenceSentence(modelId(), queryText(), ActionListener.wrap(floatList -> {
+                    vectorSetOnce.set(vectorAsListToArray(floatList));
+                    actionListener.onResponse(null);
+                }, actionListener::onFailure)))
+            );
+            return new NeuralQueryBuilder(fieldName(), queryText(), modelId(), k(), vectorSetOnce::get);
```

Comment:
[Query] : is there any reason why we are not returning "this" here and returning a new query object? and also I am not able to understand in which case line number 218 will hit?
vectorSupplier can be null or not null.

Reply:
We are not returning this here because we are changing the supplier of the query builder. If we return this, this check https://github.com/opensearch-project/OpenSearch/blob/e44158d4d10d4f8905895ffa50bf9398b8550667/server/src/main/java/org/opensearch/index/query/Rewriteable.java#L109 will see the same reference is indicate another round of rewrites does not need to be performed. So instead, we copy it. This is how Geo does it: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/query/AbstractGeometryQueryBuilder.java#L509.

> vectorSupplier can be null or not null.

Good point. Ill simplify